src/EntityFramework.Relational.Design/ReverseEngineering/CodeWriter.cs

```
@@ -0,0 +1,69 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using JetBrains.Annotations;
+using Microsoft.Data.Entity.Metadata;
+using Microsoft.Data.Entity.Relational.Design.ReverseEngineering.Configuration;
+using Microsoft.Data.Entity.Utilities;
+
+namespace Microsoft.Data.Entity.Relational.Design.ReverseEngineering
+{
+    public abstract class CodeWriter
+    {
+        private const string DefaultFileExtension = ".cs";
+
+        public CodeWriter([NotNull] IFileService fileService)
+        {
+            Check.NotNull(fileService, nameof(fileService));
+
+            FileService = fileService;
+        }
+
+        public virtual IFileService FileService { get;[param: NotNull] set; }
+        public virtual string FileExtension { get;[param: NotNull] set; } = DefaultFileExtension;
+
+        public virtual List<string> ReadOnlyOutputFiles(
```

Of the output files this class is about to write which a) already exist and b) are read-only (so that we can raise an error message saying they need to be fixed - probably by being checked out of Source Control).